src/sentry_os.c

```diff
@@ -0,0 +1,179 @@
+#include "sentry_os.h"
+#include "sentry_string.h"
+
+#ifdef SENTRY_PLATFORM_WINDOWS
+
+#    include <winver.h>
+
+sentry_value_t
+sentry__get_os_context(void)
+{
+    sentry_value_t os = sentry_value_new_object();
+    if (sentry_value_is_null(os)) {
+        return os;
+    }
+
+    sentry_value_set_by_key(os, "type", sentry_value_new_string("os"));
+    sentry_value_set_by_key(os, "name", sentry_value_new_string("Windows"));
+
+    void *ffibuf = NULL;
+
+    DWORD size = GetFileVersionInfoSizeW(L"kernel32.dll", NULL);
+    if (!size) {
+        goto fail;
+    }
+
+    ffibuf = sentry_malloc(size);
+    if (!GetFileVersionInfoW(L"kernel32.dll", 0, size, ffibuf)) {
+        goto fail;
+    }
+
+    VS_FIXEDFILEINFO *ffi;
+    UINT ffi_size;
+    if (!VerQueryValueW(ffibuf, L"\\", &ffi, &ffi_size)) {
+        goto fail;
+    }
+    ffi->dwFileFlags &= ffi->dwFileFlagsMask;
+
+    char buf[100];
+    snprintf(buf, sizeof(buf), "%u.%u.%u", ffi->dwFileVersionMS >> 16,
+        ffi->dwFileVersionMS & 0xffff, ffi->dwFileVersionLS >> 16);
+
+    sentry_value_set_by_key(os, "version", sentry_value_new_string(buf));
+
+    snprintf(buf, sizeof(buf), "%lu", ffi->dwFileVersionLS & 0xffff);
+
+    sentry_value_set_by_key(os, "build", sentry_value_new_string(buf));
+
+    sentry_free(ffibuf);
+
+    sentry_value_freeze(os);
+    return os;
+
+fail:
+    sentry_free(ffibuf);
+
+    sentry_value_decref(os);
+    return sentry_value_new_null();
+}
+
+#elif defined(SENTRY_PLATFORM_MACOS)
+
+#    include <sys/sysctl.h>
+#    include <sys/utsname.h>
+
+sentry_value_t
+sentry__get_os_context(void)
+{
+    sentry_value_t os = sentry_value_new_object();
+    if (sentry_value_is_null(os)) {
+        return os;
+    }
+
+    sentry_value_set_by_key(os, "type", sentry_value_new_string("os"));
+    sentry_value_set_by_key(os, "name", sentry_value_new_string("macOS"));
+
+    char buf[100];
+    size_t buf_len = sizeof(buf);
+
+    if (sysctlbyname("kern.osproductversion", buf, &buf_len, NULL, 0) != 0) {
+        goto fail;
+    }
+
+    size_t num_dots = 0;
+    for (size_t i = 0; i < buf_len; i++) {
+        if (buf[i] == '.') {
+            num_dots += 1;
+        }
+    }
+    if (num_dots < 2 && buf_len + 3 < sizeof(buf)) {
+        strcat(buf, ".0");
+    }
```

Comment:
I think I would opt not to do this, it puts the SDK into the business of trying to interpret what the OS wants to describe itself as and that seems like a game you can never win.

Reply:
a lot of tools, including crashpad, and our own normalization try to parse 3 semver components. also big sur is the first version that defaults to 2 components, and these things are super stable on mac at least.